DESIGN.md

```diff
@@ -73,77 +157,225 @@ Flow control support:
 
-### Behavior
+Examples:
 
-##### Creation
+- An `Observable.subscribe` passes a `Disposable` to the `Observable.onSubscribe` to allow the `Observer` to dispose of the subscription. 
+- A `Scheduler` returns a `Disposable` that you use for disposing of the `Scheduler`.
 
-Creation of a stream falls into the following use cases, all of which should be catered to in API design.
+`Subscription` is a similar type used for lifecycle management on the `Flowable` type with interactive flow control.
 
-- async, hot, push (ie. system or user events)
-- async, cold, push (ie. events resulting from remote system via network connection)
-- sync, cold, pull (ie. iterable, file, range)
-- async, cold, pull (ie. RPC/REST network call, cross-thread queue draining)
- 
-Unknown:
+##### Operator
 
-- hot, pull (what is an example of this?)
+An operator follows a specific lifecycle (union of the producer/consumer contract).
 
-Flow control support:
+- It must propagate the `subscribe` event upstream (to the producer).
+- It must obey the RxJava contract (serialize all events, `onError`/`onComplete` are terminal).
+- If it has resources to cleanup it is responsible for watching `onError`, `onComplete`, and `cancel/dispose`, and doing the necessary cleanup.
+- It must propagate the `cancel/dispose` upstream.
+
+In the addition of the previous rules, an operator for `Flowable`:
+
+- It must propagate/negotiate the `request(n)` event.
+
+
+### Creation
+
+Creation of the various types should be exposed through factory methods that provide safe construction.
+
+```java
+Flowable.create(SyncGenerator generator)
+
+Flowable.create(AsyncGenerator generator)
+
+Observable<T>.create(OnSubscribe<Observer<T>> onSubscribe)
+
+Single<T>.create(OnSubscribe<Single.Subscriber<T>> onSubscribe)
+
+Completable<T>.create(OnSubscribe<Completable.Subscriber<T>> onSubscribe)
+```
+
+### Terminal behavior
+
+A producer can terminate a stream by emitting `onComplete` or `onError`. A consumer can terminate a stream by calling `cancel`/`dispose`.
+
+Any resource cleanup of the source or operators must account for any of these three termination events. In other words, if an operator needs cleanup, then it should register the cleanup callback with `cancel`/`dispose`, `onError` and `onComplete`.
+
+The final `subscribe` will *not* invoke `cancel`/`dispose` after receiving an `onComplete` or `onError`.
+
+### JVM target and source compatibility
+
+The 2.x version will target JDK6+ but the source will be using some of the Java 8 features.
+We'll use the retrolambda project to generate JDK6 Byte-code from the Java 8 sources.
+The only Java 8 features we intend to use are: lambda expressions, method references and try-with-resources statements.
+Default methods may be considered if retrolambda support reach an acceptable level of maturity.
+
+It still up to discussion to know how we will generate multiple targets?
+Options are:
+
+- two artifcat ids: rxjava-jdk6, rxjava-jdk8
```

Comment:
Also, how would you include sources for the jdk6 distribution since retrolambda only operates on the bytecode? It would make the IDE have to decompile the classes to represent them rather than being able to just click in. Would this not make debugging unnecessarily more tedious and also sacrifice javadoc/comment access?